File: android/app/src/main/java/com/poshanforlife/android/core/network/ApiResponse.kt
package com.poshanforlife.android.core.network

import kotlinx.serialization.Serializable

/** Success envelope shape returned by the backend for every 2xx response. */
@Serializable
data class ApiResponse<T>(
    val success: Boolean,
    val data: T? = null,
    val meta: Meta? = null,
)

/** Pagination metadata, present on list endpoints only. */
@Serializable
data class Meta(
    val total: Long? = null,
    val page: Int? = null,
    val limit: Int? = null,
)

/** Error envelope shape returned by the backend for every non-2xx response. */
@Serializable
data class ApiError(
    val success: Boolean,
    val error: String,
    val code: String,
    val details: String? = null,
)

File: ios/PoshanForLife/Core/Networking/APIResponse.swift
import Foundation

/// Success envelope returned by every 2xx backend response.
/// Mirrors `backend/.../dto/ApiResponse.java` and the Android `ApiResponse.kt`
/// — this contract is shared by the Angular web app too, so it must not drift.
struct APIResponse<T: Decodable>: Decodable {
    let success: Bool
    let data: T?
    let meta: PageMeta?
}

/// Pagination metadata, present on list endpoints only.
/// The backend declares these as non-null primitives, but they're optional here
/// so a future endpoint that omits one can't fail the whole decode.
struct PageMeta: Decodable, Equatable {
    let total: Int?
    let page: Int?
    let limit: Int?
}

/// For endpoints whose useful result is "it worked" — lets them go through the
/// same generic `request` path instead of needing a separate no-content method.
struct EmptyResponse: Decodable {}
